presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,247 @@
+    {
+        Domain domain = dynamicFilterSummaries.get(SourceDescriptor.of(queryId, filterId));
+        if (domain == null) {
+            return Optional.empty();
+        }
+        else {
+            return Optional.of(domain);
+        }
+    }
+
+    public Supplier<TupleDomain<ColumnHandle>> createDynamicFilterSupplier(QueryId queryId, List<DynamicFilters.Descriptor> dynamicFilters, Map<Symbol, ColumnHandle> columnHandles)
+    {
+        Map<String, ColumnHandle> sourceColumnHandles = extractSourceColumnHandles(dynamicFilters, columnHandles);
+
+        return () -> dynamicFilters.stream()
+                .map(filter -> getSummary(queryId, filter.getId())
+                        .map(summary -> translateSummaryToTupleDomain(filter.getId(), summary, sourceColumnHandles)))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .reduce(TupleDomain.all(), TupleDomain::intersect);
+    }
+
+    @Immutable
+    private static class SourceDescriptor
+    {
+        private final QueryId queryId;
+        private final String filterId;
+
+        public static SourceDescriptor of(QueryId queryId, String filterId)
+        {
+            return new SourceDescriptor(queryId, filterId);
+        }
+
+        private SourceDescriptor(QueryId queryId, String filterId)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.filterId = requireNonNull(filterId, "filterId is null");
+        }
+
+        public QueryId getQueryId()
+        {
+            return queryId;
+        }
+
+        public String getFilterId()
+        {
+            return filterId;
+        }
+
+        @Override
+        public boolean equals(Object other)
+        {
+            if (other == this) {
+                return true;
+            }
+            if (other == null || getClass() != other.getClass()) {
+                return false;
+            }
+
+            SourceDescriptor sourceDescriptor = (SourceDescriptor) other;
+
+            return Objects.equals(queryId, sourceDescriptor.queryId) &&
+                    Objects.equals(filterId, sourceDescriptor.filterId);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(queryId, filterId);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("queryId", queryId)
+                    .add("filterId", filterId)
+                    .toString();
+        }
+    }
+
+    private static TupleDomain<ColumnHandle> translateSummaryToTupleDomain(String filterId, Domain summary, Map<String, ColumnHandle> sourceColumnHandles)
+    {
+        if (summary.isNone()) {
+            return TupleDomain.none();
+        }
+        ColumnHandle sourceColumnHandle = requireNonNull(sourceColumnHandles.get(filterId), String.format("Source column handle for dynamic filter %s is null", filterId));
```

Comment:
replace with non-template string or use supplier version of `requireNonNull`. Otherwise message string will be formatted in every `translateSummaryToTupleDomain` call

Reply:
done